### DISCUSSION TITLE
What does it take to get HTTP/3 out of Experimental state?

### DISCUSSION THREAD

**Discussion by nibanks** [curl/curl#10156] [Q&A] [answered]
[HTTP3.md](https://github.com/curl/curl/blob/72b819f14792f350e90a7b7341d41701e36c0557/docs/HTTP3.md) states:

> HTTP/3 and QUIC support in curl is considered EXPERIMENTAL until further notice.

Is there a checklist or something, somewhere that should be met before moving out of experimental state? I'd be very interested/motivated in pushing to get `msh3` to that state.

**Comment by bagder** [comment]
Yes, I would like to get such a checklist done so that we can work on making sure we tick off all the boxes correctly.

@icing has a "h3 work for curl" project in his lap so let's get together and work out a plan.

**Comment by dfandrich** [comment]
There are a few items in KNOWN_BUGS, most of which would likely end up on that list.

**Reply by nibanks** [reply]
Copying those here:

 > 18. HTTP/3
 > 18.1 If the HTTP/3 server closes connection during upload curl hangs
 > 18.2 Transfer closed with n bytes remaining to read
 > 18.4 timeout when reusing an http3 connection
 > 18.9 connection migration does not work

* Does anyone know which HTTP/3 implementation was used for these?
* For 18.4, are there more details? The connection is obviously expected to timeout after a certain period of time with no use.
* For 18.9, what exactly is the criteria for it to "work". Does the client need to be able to switch local address/port, or does the connection just need to survive NAT rebinding?

**Reply by bagder** [reply]
If you click through to the original issues, they usually include more details.

I think more generally, I mean beyond some individual bugs we might still have, we need:

1. HTTP/3 test cases that verify basic operations, multiplexing, go away etc.
2. A fallback solution to non-h3 when QUIC/h3 fails

@icing is set out to work on these items and hopefully we will see some results from this soonish.

Then, as soon as we think it feels stable enough and we are happy with the API to set the HTTP/3 selection etc, I think we can remove the experimental tag. There is no formal procedure for this but I think we will keep the discussion and development alive and progressing until we agree we reach that point.

**Reply by icing** [reply]
I think we can only make progress on these issues if we have a test suite to reproduce them. That is high on my list.

**Comment by bagder** [comment, marked as answer]
The HTTP/3 page on the curl site now contains a section for [what it takes](https://curl.se/docs/http3.html#Experimental).
